### ISSUE TITLE
Persistent post database.

### ISSUE THREAD

**Issue by e5150** [instaloader/instaloader#72]
A possible solution to issues like the localtime vs. UTC thing might be to keep a database of all images that have been downloaded. I'm thinking that the user and the file's basename (as on the IG servers) can be used. (Assuming the filenames are at least locally unique with regards to their owner.) If each file was stored as `user/22382_5984....jpg` and hard-linked to whatever `filename-pattern` for the post is. Then the file would not have to be re-downloaded if `filename-pattern` is changed by the user, or Instaloader changes its behaviour (or even if the pattern differs because the post is requested by its hashtag, if it was previously downloaded from feed or profile).
With a logic similar to:
```
def download(url, user, dest): # dest based on pattern
        raw_name = os.path.join(db_root, user, os.path.basename(url))
        if os.path.isfile(dest):
                if not os.path.isfile(raw_name):
                        os.link(dest, raw_name)
                return False
        if os.path.isfile(raw_name):
                if not os.path.isfile(dest):
                        os.link(raw_name, dest)
                return False
        actually_download(url, dest)
        os.link(dst, raw_name)
        return True
```
User's numeric ID could be used to really make it persistent, without having to rename if the username changes. But I don't know if non-unix systems have something like hard-links, IIRC Windows introduced something akin to symlinks around Vista.
Symbolic links could also be used, but in either case there would be an issue of redundant files if either the "raw" or "pattern" file was removed. Residual files taking up space and dangling symlinks, for hard and symbolic links, respectively. (Which could easily be solved by `find download_dir -name '*.jpg' -links 1 -delete` or `find -L download_dir -type l -delete`. But that might be an excessive burden on the user.) Another way could be to just store the information in a `user-id.json` file containing a `Dict[str, List[str]]` of `{ "raw-name.jpg" : ["stored-as-pattern-A.jpg", "stored-as-pattern-B.jpg"] }`, where the files might as well be hard-linked, but only to one-another, and no raw-name.jpg would appear on the file system.

Any thoughts?

**Comment by raphaelbernardino** [comment]
I guess you have all figured out by now, but why can't you save [timestamp](https://en.wikipedia.org/wiki/Unix_time) from last successful run (or update)? You don't have to store all images (or hashes/links) to `--fast-update` work, right?

You can compare the last run timestamp with current timestamp and download required missing data (photos/videos/stories and whatever). If so, database would keep track of scraped users id and timestamp only. If i am not wrong, you can also extract id from a post. I don't know if it works for stories too.

In the last commit https://github.com/instaloader/instaloader/commit/903d882015f30148b6a1e3d5c6aea470f039ae56 seems like you download data, calculate hash and then checks if its on database. That might have some impact on Instagram limits.

**Comment by e5150** [comment]
I'm inclined to agree with @raphaelbernardino - the main point is to avoid having to re-download old images with a new pattern.

>seems like you download data, calculate hash and then checks if its on database. That might have some impact on Instagram limits.

By reading `903d882` I don't think there would be a difference in the number of requests with a timestamp database. Both approaches would request a set of data (shortcode, timestamp, url etc.) for some posts. The current code checks the hash of the old file on the filesystem, and (if it matches what was stored when it was downloaded) links it to the new filename.

As far as I understand there is no limit on the number of images downloaded, just GQL-queries. It still saves some bandwidth and storage space to not have to re-download a `#cutecat` post that already occured in my `:feed` by just hardlinking them together. But it would probably be much easier (and non-unix compatible) to implement a db-table consisting only of `target` and `utc-timestamp`. And accept the extra bandwidth cost. Another plus for the timestamp db is that I could delete old images from my filesystem, if they aren't needed for `--fast-update`.
There are also some issues with the current approach, storing `os.path.abspath(file)` will make the db useless if I move the entire tree. And if I've just downloaded a profile I'm following, some recent post might make `--fast-update :feed` stop prematurely.  The db also violates [1NF](https://en.wikipedia.org/wiki/First_normal_form), a pet-peeve of mine. :)

If one just wants to rename old files according to a new pattern, an external script could be used to parse the stored json files.

tl;dr: A timestamp database solves the main issue (relying on file existance from an old pattern), works on windows and would probably be overall simpler.

**Comment by aandergr** [comment]
> If one just wants to rename old files according to a new pattern, an external script could be used to parse the stored json files.

I drafted an external script to parse the stored json files and mass rename according to a new pattern:
https://gist.github.com/aandergr/69e08367d6b9632375691dda35eb649c
